source/locales.tex

```diff
@@ -4632,3 +4651,719 @@
+\end{itemize}
+\end{itemdescr}
+
+\indexlibraryctor{text_encoding}%
+\begin{itemdecl}
+constexpr text_encoding(id i) noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\tcode{i} has the value of one of the enumerators of \tcode{id}.
+
+\pnum
+\ensures
+\begin{itemize}
+\item
+\tcode{\exposid{mib_} == i} is \tcode{true}.
+\item
+If \tcode{(\exposid{mib_} == id::unknown || \exposid{mib_} == id::other)}
+is \tcode{true},
+\tcode{strlen(\exposid{name_}) == 0} is \tcode{true}.
+Otherwise,
+\tcode{ranges::contains(aliases(), string_view(\exposid{name_}))}
+is \tcode{true}.
+\end{itemize}
+\end{itemdescr}
+
+\indexlibrarymember{mib}{text_encoding}%
+\begin{itemdecl}
+constexpr id mib() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{\exposid{mib_}}.
+\end{itemdescr}
+
+\indexlibrarymember{name}{text_encoding}%
+\begin{itemdecl}
+constexpr const char* name() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{\exposid{name_}} if \tcode{(\exposid{name_}[0] != '\textbackslash 0')}
+is \tcode{true},
+\keyword{nullptr} otherwise;
+
+\pnum
+\remarks
+If \tcode{name() == nullptr} is \tcode{false},
+\tcode{name()} is an NTBS and
+accessing elements of \tcode{\exposid{name_}}
+%FIXME: This isn't a range.
+outside of the range \tcode{name() +} \range{0}{strlen(name()) + 1}
+is undefined behavior.
+\end{itemdescr}
+
+\indexlibrarymember{aliases}{text_encoding}%
+\begin{itemdecl}
+constexpr aliases_view aliases() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+Let \tcode{r} denote an instance of \tcode{aliases_view}.
+If \tcode{*this} represents a known registered character encoding then:
+\begin{itemize}
+\item
+\tcode{r.front()} is the primary name of the registered character encoding,
+\item
+\tcode{r} contains the aliases of the registered character encoding, and
+\item
+\tcode{r} does not contain duplicate values when compared with \tcode{strcmp}.
+\end{itemize}
+Otherwise, \tcode{r} is an empty range.
+
+\pnum
+Each element in \tcode{r}
+is a non-null, non-empty NTBS encoded in the literal character encoding and
+comprising only characters from the basic character set.
+
+\pnum
+\returns
+\tcode{r}.
+
+\pnum
+\begin{note}
+The order of aliases in \tcode{r} is unspecified.
+\end{note}
+\end{itemdescr}
+
+\indexlibrarymember{literal}{text_encoding}%
+\begin{itemdecl}
+static consteval text_encoding literal() noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\mandates
+\tcode{CHAR_BIT == 8} is \tcode{true}.
+
+\pnum
+\returns
+A \tcode{text_encoding} object representing
+the ordinary character literal encoding\iref{lex.charset}.
+\end{itemdescr}
+
+\indexlibrarymember{environment}{text_encoding}%
+\begin{itemdecl}
+static text_encoding environment();
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\mandates
+\tcode{CHAR_BIT == 8} is \tcode{true}.
+
+\pnum
+\returns
+A \tcode{text_encoding} object representing
+the implementation-defined character encoding scheme of the environment.
```

Comment:
use \impldef macro

Reply:
Same issue here - see uses of \impldef in the locations where the determination is stated.

Reply:
Do we say elsewhere that this property is implementation-defined? If not, maybe it would be good to have this in the index via \impldef.

Reply:
Yes, I believe we do. Here's my interpretation:

```
+\pnum
+The set of aliases of a known registered character encoding is an
+\impldef{set of aliases of a known registered character encoding}
+superset of the aliases specified in the IANA Character Sets registry.
```

Here we're saying the encoding scheme is \impldef.

```
+\pnum
+How a \tcode{text_encoding} object
+is determined to be representative of a character encoding scheme
+implemented in the translation or execution environment is
+\impldef{how \tcode{text_encoding} objects are
+determined to be representative of a character encoding scheme}.
```

Here we're saying the text_encoding representation of the encoding
scheme is \impldef.

Reply:
Yes, that's all fine, but we have a third kind of impl-def here:  What the (abstract) character encoding scheme of the environment actually is.

Reply:
I changed to `\impldef{character encoding scheme of the environment}`, but I think Johel's last suggestion would be the correct fix (but more than editorial):
>> or such a property is defined (and indexed) elsewhere
>> and the following sentence should have been a note

What about the other "implementation-defined"? Are we OK interpreting that as a use?

Reply:
I think the other occurrences are all either in notes or pre-existing except for the _Returns:_ element of `locale::encoding`. I think we may be missing some normative wording that says that locales are associated with encoding schemes; perhaps we should handle this as an issue.

Reply:
Deferred to issue.